packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -2,347 +2,196 @@
     }
-    final CharacterRange range = CharacterRange.at(_text, startOffset, endOffset);
-    return TextPosition(
-      offset: _text.length - range.stringAfterLength,
-      affinity: TextAffinity.upstream,
-    );
+    final CharacterRange rangeAtPosition = CharacterRange.at(_text, position);
+    return rangeAtPosition.isNotEmpty
+      ? TextRange(start: rangeAtPosition.stringBeforeLength, end: rangeAtPosition.stringBeforeLength + rangeAtPosition.current.length)
+      // rangeAtPosition is empty means `position` is a grapheme boundary.
+      : TextRange(start: rangeAtPosition.stringBeforeLength, end: getTrailingTextBoundaryAt(position) ?? -1);
   }
 }
 
-/// A text boundary that uses words as logical boundaries.
+/// A [TextBoundary] subclass for locating word breaks.
 ///
-/// This class uses [UAX #29](https://unicode.org/reports/tr29/) defined word
-/// boundaries to calculate its logical boundaries.
+/// The underlying implementation uses [UAX #29](https://unicode.org/reports/tr29/)
+/// defined default word boundaries.
+///
+/// The default word break rules keep horizontal whitespaces together as a
+/// single word. For instance, "hello world" contains 3 words, which is usually
+/// not what the user expects. Consider using the [until] method to skip
+/// unwanted words.
 class WordBoundary extends TextBoundary {
   /// Creates a [WordBoundary] with the text and layout information.
   const WordBoundary(this._textLayout);
 
   final TextLayoutMetrics _textLayout;
 
   @override
-  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
-    return TextPosition(
-      offset: _textLayout.getWordBoundary(position).start,
-      affinity: TextAffinity.downstream,  // ignore: avoid_redundant_argument_values
-    );
-  }
-
-  @override
-  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
-    return TextPosition(
-      offset: _textLayout.getWordBoundary(position).end,
-      affinity: TextAffinity.upstream,
-    );
-  }
+  TextRange getTextBoundaryAt(int position) => _textLayout.getWordBoundary(TextPosition(offset: max(position, 0)));
 
-  @override
-  TextRange getTextBoundaryAt(TextPosition position) {
-    return _textLayout.getWordBoundary(position);
-  }
+  /// Creates a new [TextBoundary] by repeatedly expanding a [WordBoundary]
+  /// until the given `predicate` returns true.
+  ///
+  /// The predicate will be evaluated every time the current boundary is
+  /// expanded. The `offset` in the predicate is the code unit offset in the
+  /// text that represents the new boundary location, and `forward` indicates
+  /// the direction of the expansion. For example, to skip word boundaries that
+  /// are adjacent to whitespaces:
+  ///
+  /// ```dart
+  /// WordBoundary(textLayout)
+  ///   .until((int offset, bool forward) => !isWhitespace(text.codeUnitAt(forward ? offset - 1 : offset)));
+  /// ```
+  TextBoundary until(UntilPredicate predicate) => _UntilTextBoundary(this, predicate);
```

Comment:
Can you think of use case other than skipping whitespace? maybe just hard code a `skipWhiteSpace` method so that you don't need to duplicate code on both renderparagraph and rendereditable? WDYT

Reply:
unfortunately this can't be done right now,  WordBoundary itself doesn't have access to code units

Reply:
Do you think it worth adding access to the code unit for word boundary?

Reply:
I can move this class to TextPainter so we have random access to the code units